File: android/src/main/kotlin/com/example/flutter_webrtc_noise_suppressor/FlutterWebrtcNoiseSuppressorPlugin.kt
package com.example.flutter_webrtc_noise_suppressor

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlin.math.exp
import kotlin.math.sqrt

// ---------------------------------------------------------------------------
// RMS Gate Processor
//
// Implements the same algorithm as the C++ NoiseSuppressorProcessor:
//   1. Compute RMS energy over the frame.
//   2. Compare to threshold; manage a hold countdown.
//   3. Apply IIR gain smoothing (separate attack / release time constants).
//   4. Multiply every sample by the smoothed gain.
//
// Short values are in the range [-32768, 32767]. We scale to [-1, 1] for the
// RMS computation, then scale the gain factor back to the integer domain for
// the output multiply.
// ---------------------------------------------------------------------------
class RmsGateProcessor {

    // Configuration (written from the main thread, read on the audio thread).
    // Kotlin does not have std::atomic, but @Volatile is sufficient here
    // because we only ever write a single word and can tolerate one stale read.
    @Volatile var threshold: Float = 0.02f
    @Volatile var holdMs: Int = 200
    @Volatile var residualGain: Float = 0.05f
    @Volatile var attackMs: Float = 5.0f
    @Volatile var releaseMs: Float = 80.0f

    // Last computed RMS level — written on the audio thread, readable from any thread.
    @Volatile var lastRms: Float = 0.0f

    // Audio-thread state (touched only in initialize / process / reset).
    private var sampleRateHz: Int = 0
    private var channels: Int = 0
    private var samplesPerFrame: Int = 0
    private var holdFramesRemaining: Int = 0
    private var smoothedGain: Float = 0.0f
    private var attackCoeff: Float = 0.0f
    private var releaseCoeff: Float = 0.0f

    fun initialize(sampleRateHz: Int, channels: Int) {
        this.sampleRateHz = sampleRateHz
        this.channels = channels
        // WebRTC processes in 10 ms chunks.
        this.samplesPerFrame = sampleRateHz / 100
        holdFramesRemaining = 0
        smoothedGain = 0.0f
        recomputeCoefficients()
    }

    fun process(audioBuffer: ShortArray) {
        if (sampleRateHz <= 0 || samplesPerFrame <= 0) return

        val n = audioBuffer.size
        if (n == 0) return

        val thresholdLocal = threshold
        val residualLocal = residualGain

        // Snapshot of hold_ms for this frame (safe to read once).
        val holdMsLocal = holdMs

        // ------------------------------------------------------------------
        // Compute RMS — scale from 16-bit integers to [-1, 1].
        // ------------------------------------------------------------------
        val scale = 1.0f / 32768.0f
        var sumSq = 0.0f
        for (sample in audioBuffer) {
            val f = sample.toFloat() * scale
            sumSq += f * f
        }
        val rms = sqrt(sumSq / n.toFloat())
        lastRms = rms

        // ------------------------------------------------------------------
        // Gate logic with hold
        // ------------------------------------------------------------------
        val gateOpen: Boolean
        if (rms >= thresholdLocal) {
            val holdFrames =
                if (sampleRateHz > 0)
                    (holdMsLocal.toFloat() * sampleRateHz.toFloat() / 1000.0f /
                            samplesPerFrame.toFloat()).toInt()
                else 0
            holdFramesRemaining = holdFrames
            gateOpen = true
        } else if (holdFramesRemaining > 0) {
            holdFramesRemaining--
            gateOpen = true
        } else {
            gateOpen = false
        }

        // ------------------------------------------------------------------
        // IIR gain smoothing
        // ------------------------------------------------------------------
        val targetGain = if (gateOpen) 1.0f else residualLocal

        smoothedGain = if (targetGain > smoothedGain) {
            // Opening — use attack coefficient.
            attackCoeff * smoothedGain + (1.0f - attackCoeff) * targetGain
        } else {
            // Closing — use release coefficient.
            releaseCoeff * smoothedGain + (1.0f - releaseCoeff) * targetGain
        }
        smoothedGain = smoothedGain.coerceIn(0.0f, 1.0f)

        // ------------------------------------------------------------------
        // Apply gain in-place (ShortArray — scale gain to integer multiply).
        // Clamp output to Short range to avoid overflow.
        // ------------------------------------------------------------------
        for (i in audioBuffer.indices) {
            val scaled = audioBuffer[i].toFloat() * smoothedGain
            audioBuffer[i] = scaled.toInt().toShort()
        }
    }

    fun reset(newRate: Int) {
        sampleRateHz = newRate
        samplesPerFrame = newRate / 100
        holdFramesRemaining = 0
        smoothedGain = 0.0f
        recomputeCoefficients()
    }

    fun release() {
        sampleRateHz = 0
        channels = 0
        samplesPerFrame = 0
        holdFramesRemaining = 0
        smoothedGain = 0.0f
        attackCoeff = 0.0f
        releaseCoeff = 0.0f
    }

    // -----------------------------------------------------------------------
    // Private helpers
    // -----------------------------------------------------------------------

    private fun recomputeCoefficients() {
        if (sampleRateHz <= 0 || samplesPerFrame <= 0) {
            attackCoeff = 0.0f
            releaseCoeff = 0.0f
            return
        }
        val spf = samplesPerFrame.toFloat()
        val sr = sampleRateHz.toFloat()
        val atk = attackMs
        val rel = releaseMs

        // coeff = exp(-samples_per_frame / (time_ms * sample_rate_hz / 1000))
        attackCoeff = if (atk > 0.0f) exp(-spf / (atk * sr / 1000.0f)) else 0.0f
        releaseCoeff = if (rel > 0.0f) exp(-spf / (rel * sr / 1000.0f)) else 0.0f
    }
}

// ---------------------------------------------------------------------------
// Processing mode constants (must match Dart-side NoiseProcessingMode.index)
// ---------------------------------------------------------------------------
private const val MODE_DISABLED = 0
private const val MODE_RMS_GATE = 1
private const val MODE_VAD_GATE = 2  // falls back to rmsGate
private const val MODE_DENOISE  = 3  // falls back to disabled (future)
private const val MODE_HYBRID   = 4  // falls back to disabled (future)

// ---------------------------------------------------------------------------
// Flutter Plugin
// ---------------------------------------------------------------------------
class FlutterWebrtcNoiseSuppressorPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var processor: RmsGateProcessor? = null

    // Current mode (main thread only — method channel calls are serialised).
    private var currentMode: Int = MODE_DISABLED

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            binding.binaryMessenger,
            "flutter_webrtc_noise_suppressor"
        )
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        safeDisposeProcessor()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize"    -> handleInitialize(result)
            "dispose"       -> handleDispose(result)
            "setMode"       -> handleSetMode(call, result)
            "configure"     -> handleConfigure(call, result)
            "getAudioLevel" -> result.success(processor?.lastRms?.toDouble() ?: 0.0)
            else            -> result.notImplemented()
        }
    }

    // -----------------------------------------------------------------------
    // Method handlers
    // -----------------------------------------------------------------------

    private fun handleInitialize(result: Result) {
        if (processor == null) {
            processor = RmsGateProcessor()
        }

        // Register with flutter_webrtc's audio processing pipeline.
        //
        // flutter_webrtc exposes AudioProcessingController with a
        // capturePostProcessing property that accepts an
        // ExternalAudioFrameProcessing implementation.
        //
        // We use reflection so that this plugin compiles and loads even when
        // flutter_webrtc is not present in the host app (the plugin will
        // simply not process audio in that case).
        try {
            val fwrtcClass = Class.forName(
                "com.cloudwebrtc.webrtc.FlutterWebRTCPlugin"
            )
            val getController = fwrtcClass.getMethod(
                "getAudioProcessingController"
            )
            // FlutterWebRTCPlugin is a singleton — retrieve the instance via
            // its static getInstance() method.
            val getInstance = fwrtcClass.getMethod("getInstance")
            val pluginInstance = getInstance.invoke(null)
            val controller = getController.invoke(pluginInstance)

            if (controller != null) {
                val capturePost = controller.javaClass
                    .getField("capturePostProcessing")
                    .get(controller)

                if (capturePost != null) {
                    val addProcessor = capturePost.javaClass.getMethod(
                        "addProcessor",
                        Class.forName(
                            "com.cloudwebrtc.webrtc.audio.ExternalAudioFrameProcessing"
                        )
                    )
                    addProcessor.invoke(capturePost, buildProcessingProxy())
                }
            }
        } catch (e: Exception) {
            // flutter_webrtc is not present or API has changed.
            // Log and continue — the processor is still instantiated and can
            // be used if the caller wires it up manually.
            android.util.Log.w(
                "NoiseSuppressor",
                "Could not register with flutter_webrtc AudioProcessingController: ${e.message}"
            )
        }

        result.success(null)
    }

    private fun handleDispose(result: Result) {
        if (processor == null) {
            result.error("NOT_INITIALIZED", "dispose() called before initialize()", null)
            return
        }

        // Unregister from flutter_webrtc (mirror of handleInitialize).
        try {
            val fwrtcClass = Class.forName(
                "com.cloudwebrtc.webrtc.FlutterWebRTCPlugin"
            )
            val getInstance = fwrtcClass.getMethod("getInstance")
            val pluginInstance = getInstance.invoke(null)
            val getController = fwrtcClass.getMethod("getAudioProcessingController")
            val controller = getController.invoke(pluginInstance)

            if (controller != null) {
                val capturePost = controller.javaClass
                    .getField("capturePostProcessing")
                    .get(controller)

                if (capturePost != null) {
                    // Remove all processors registered by this plugin.
                    // flutter_webrtc's API may provide removeAllProcessors()
                    // or removeProcessor(processor). Try both.
                    try {
                        val removeAll = capturePost.javaClass
                            .getMethod("removeAllProcessors")
                        removeAll.invoke(capturePost)
                    } catch (_: NoSuchMethodException) {
                        // removeAllProcessors not available; ignore.
                    }
                }
            }
        } catch (e: Exception) {
            android.util.Log.w(
                "NoiseSuppressor",
                "Could not unregister from flutter_webrtc: ${e.message}"
            )
        }

        safeDisposeProcessor()
        result.success(null)
    }

    private fun handleSetMode(call: MethodCall, result: Result) {
        if (processor == null) {
            result.error("NOT_INITIALIZED", "setMode() called before initialize()", null)
            return
        }
        val mode = call.argument<Int>("mode")
        if (mode == null) {
            result.error("INVALID_ARGUMENT", "missing required argument: mode", null)
            return
        }
        currentMode = mode
        result.success(null)
    }

    private fun handleConfigure(call: MethodCall, result: Result) {
        val p = processor
        if (p == null) {
            result.error("NOT_INITIALIZED", "configure() called before initialize()", null)
            return
        }

        call.argument<Double>("threshold")?.let { p.threshold = it.toFloat() }
        call.argument<Int>("holdMs")?.let { p.holdMs = it }
        call.argument<Double>("residualGain")?.let { p.residualGain = it.toFloat() }
        call.argument<Double>("attackMs")?.let { p.attackMs = it.toFloat() }
        call.argument<Double>("releaseMs")?.let { p.releaseMs = it.toFloat() }
        // vadThreshold is accepted but reserved for future VAD mode (no-op in Phase 1).

        result.success(null)
    }

    // -----------------------------------------------------------------------
    // Build a dynamic proxy that adapts RmsGateProcessor to flutter_webrtc's
    // ExternalAudioFrameProcessing interface without a compile-time dep.
    // -----------------------------------------------------------------------
    private fun buildProcessingProxy(): Any {
        val processorRef = processor!!
        val modeRef get() = currentMode

        return java.lang.reflect.Proxy.newProxyInstance(
            javaClass.classLoader,
            arrayOf(
                Class.forName(
                    "com.cloudwebrtc.webrtc.audio.ExternalAudioFrameProcessing"
                )
            )
        ) { _, method, args ->
            when (method.name) {
                "initialize" -> {
                    val sampleRate = args?.getOrNull(0) as? Int ?: return@newProxyInstance null
                    val channels   = args?.getOrNull(1) as? Int ?: return@newProxyInstance null
                    processorRef.initialize(sampleRate, channels)
                    null
                }
                "process" -> {
                    val buffer = args?.getOrNull(0) as? ShortArray
                        ?: return@newProxyInstance null
                    // Only apply processing for active modes.
                    // vadGate falls back to rmsGate; denoise/hybrid pass through.
                    when (modeRef) {
                        MODE_RMS_GATE, MODE_VAD_GATE -> processorRef.process(buffer)
                        // TODO(phase-2): MODE_DENOISE, MODE_HYBRID — run ONNX inference.
                        else -> { /* passthrough */ }
                    }
                    null
                }
                "reset" -> {
                    val newRate = args?.getOrNull(0) as? Int ?: return@newProxyInstance null
                    processorRef.reset(newRate)
                    null
                }
                "release" -> {
                    processorRef.release()
                    null
                }
                else -> null
            }
        }
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private fun safeDisposeProcessor() {
        processor?.release()
        processor = null
    }
}

File: ios/Classes/FlutterWebrtcNoiseSuppressorPlugin.m
#import "FlutterWebrtcNoiseSuppressorPlugin.h"

#import <math.h>

// ---------------------------------------------------------------------------
// Processing mode constants — must match Dart-side NoiseProcessingMode.index.
// ---------------------------------------------------------------------------
typedef NS_ENUM(NSInteger, NoiseProcessingMode) {
    NoiseProcessingModeDisabled = 0,
    NoiseProcessingModeRmsGate  = 1,
    NoiseProcessingModeVadGate  = 2,  // falls back to rmsGate
    NoiseProcessingModeDenoise  = 3,  // falls back to disabled (future)
    NoiseProcessingModeHybrid   = 4,  // falls back to disabled (future)
};

// ---------------------------------------------------------------------------
// Default configuration values
// ---------------------------------------------------------------------------
static const float kDefaultThreshold    = 0.02f;
static const int   kDefaultHoldMs       = 200;
static const float kDefaultResidualGain = 0.05f;
static const float kDefaultAttackMs     = 5.0f;
static const float kDefaultReleaseMs    = 80.0f;

// ---------------------------------------------------------------------------
// FlutterWebrtcNoiseSuppressorPlugin
//
// This class implements both FlutterPlugin (for the method channel) and the
// ExternalAudioProcessingDelegate protocol from flutter_webrtc (for the audio
// callbacks). We use NSProtocolFromString / respondsToSelector so that the
// plugin compiles and loads even when flutter_webrtc is not linked — the
// audio hooks simply won't fire in that case.
// ---------------------------------------------------------------------------
@implementation FlutterWebrtcNoiseSuppressorPlugin {
    // -----------------------------------------------------------------------
    // Configuration — written from the main thread, read on the audio thread.
    // We use @synchronized(self) for the scalar ivars because iOS does not
    // expose C11 _Atomic in ObjC conveniently. Alternatively NSLock could
    // be used, but @synchronized has negligible overhead for infrequent
    // UI-thread writes.
    // -----------------------------------------------------------------------
    float   _threshold;
    int     _holdMs;
    float   _residualGain;
    float   _attackMs;
    float   _releaseMs;
    NoiseProcessingMode _mode;

    // -----------------------------------------------------------------------
    // Audio-thread state — only touched in the delegate callbacks (serialised
    // by the WebRTC audio thread).
    // -----------------------------------------------------------------------
    double  _sampleRateHz;
    int     _numChannels;
    int     _samplesPerFrame;
    int     _holdFramesRemaining;
    float   _smoothedGain;
    float   _attackCoeff;
    float   _releaseCoeff;

    // Retain the proxy object we registered so we can remove it later.
    id      _processingProxy;

    // Last computed RMS level — written on audio thread, read on main thread.
    // Protected by @synchronized(self) for consistency with other ivars.
    float   _rmsLevel;
}

// ---------------------------------------------------------------------------
// FlutterPlugin registration
// ---------------------------------------------------------------------------

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel =
        [FlutterMethodChannel methodChannelWithName:@"flutter_webrtc_noise_suppressor"
                                   binaryMessenger:[registrar messenger]];
    FlutterWebrtcNoiseSuppressorPlugin* instance =
        [[FlutterWebrtcNoiseSuppressorPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _threshold    = kDefaultThreshold;
        _holdMs       = kDefaultHoldMs;
        _residualGain = kDefaultResidualGain;
        _attackMs     = kDefaultAttackMs;
        _releaseMs    = kDefaultReleaseMs;
        _mode         = NoiseProcessingModeDisabled;

        _sampleRateHz         = 0;
        _numChannels          = 0;
        _samplesPerFrame      = 0;
        _holdFramesRemaining  = 0;
        _smoothedGain         = 0.0f;
        _attackCoeff          = 0.0f;
        _releaseCoeff         = 0.0f;
    }
    return self;
}

// ---------------------------------------------------------------------------
// FlutterPlugin method channel handler
// ---------------------------------------------------------------------------

- (void)handleMethodCall:(FlutterMethodCall*)call
                  result:(FlutterResult)result {
    if ([call.method isEqualToString:@"initialize"]) {
        [self handleInitialize:result];
    } else if ([call.method isEqualToString:@"dispose"]) {
        [self handleDispose:result];
    } else if ([call.method isEqualToString:@"setMode"]) {
        [self handleSetMode:call result:result];
    } else if ([call.method isEqualToString:@"configure"]) {
        [self handleConfigure:call result:result];
    } else if ([call.method isEqualToString:@"getAudioLevel"]) {
        float level;
        @synchronized(self) { level = _rmsLevel; }
        result(@(level));
    } else {
        result(FlutterMethodNotImplemented);
    }
}

// ---------------------------------------------------------------------------
// initialize
// ---------------------------------------------------------------------------

- (void)handleInitialize:(FlutterResult)result {
    // Locate AudioManager.sharedInstance via the Objective-C runtime so that
    // we do not have a hard compile-time dependency on flutter_webrtc.
    Class audioManagerClass = NSClassFromString(@"AudioManager");
    if (!audioManagerClass) {
        NSLog(@"[NoiseSuppressor] AudioManager class not found — "
              "is flutter_webrtc linked?");
        result(nil);
        return;
    }

    id audioManager = [audioManagerClass performSelector:@selector(sharedInstance)];
    if (!audioManager) {
        NSLog(@"[NoiseSuppressor] AudioManager.sharedInstance returned nil");
        result(nil);
        return;
    }

    // Get capturePostProcessingAdapter.
    SEL captureAdapterSel = NSSelectorFromString(@"capturePostProcessingAdapter");
    if (![audioManager respondsToSelector:captureAdapterSel]) {
        NSLog(@"[NoiseSuppressor] capturePostProcessingAdapter not available");
        result(nil);
        return;
    }

    id adapter =
        [audioManager performSelector:captureAdapterSel];
    if (!adapter) {
        NSLog(@"[NoiseSuppressor] capturePostProcessingAdapter is nil");
        result(nil);
        return;
    }

    // Add self as a processing delegate.
    SEL addProcessingSel = NSSelectorFromString(@"addProcessing:");
    if (![adapter respondsToSelector:addProcessingSel]) {
        NSLog(@"[NoiseSuppressor] addProcessing: not available on adapter");
        result(nil);
        return;
    }

    [adapter performSelector:addProcessingSel withObject:self];
    _processingProxy = adapter;  // retain for later removal

    result(nil);
}

// ---------------------------------------------------------------------------
// dispose
// ---------------------------------------------------------------------------

- (void)handleDispose:(FlutterResult)result {
    if (_processingProxy) {
        SEL removeProcessingSel = NSSelectorFromString(@"removeProcessing:");
        if ([_processingProxy respondsToSelector:removeProcessingSel]) {
            [_processingProxy performSelector:removeProcessingSel
                                   withObject:self];
        }
        _processingProxy = nil;
    }

    // Reset audio-thread state.
    _sampleRateHz        = 0;
    _numChannels         = 0;
    _samplesPerFrame     = 0;
    _holdFramesRemaining = 0;
    _smoothedGain        = 0.0f;
    _attackCoeff         = 0.0f;
    _releaseCoeff        = 0.0f;

    result(nil);
}

// ---------------------------------------------------------------------------
// setMode
// ---------------------------------------------------------------------------

- (void)handleSetMode:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* args = call.arguments;
    NSNumber* modeNum = args[@"mode"];
    if (!modeNum) {
        result([FlutterError errorWithCode:@"INVALID_ARGUMENT"
                                   message:@"missing required argument: mode"
                                   details:nil]);
        return;
    }
    @synchronized(self) {
        _mode = (NoiseProcessingMode)[modeNum integerValue];
    }
    result(nil);
}

// ---------------------------------------------------------------------------
// configure
// ---------------------------------------------------------------------------

- (void)handleConfigure:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* args = call.arguments;

    @synchronized(self) {
        NSNumber* v;
        if ((v = args[@"threshold"]))    _threshold    = [v floatValue];
        if ((v = args[@"holdMs"]))       _holdMs       = [v intValue];
        if ((v = args[@"residualGain"])) _residualGain = [v floatValue];
        if ((v = args[@"attackMs"]))     _attackMs     = [v floatValue];
        if ((v = args[@"releaseMs"]))    _releaseMs    = [v floatValue];
        // vadThreshold is accepted but reserved for future VAD mode (no-op in Phase 1).
    }

    result(nil);
}

// ---------------------------------------------------------------------------
// ExternalAudioProcessingDelegate — called by flutter_webrtc on the audio
// thread.
//
// We conform to the protocol dynamically; the selectors must match exactly
// what flutter_webrtc's RTCAudioProcessingAdapter calls.
// ---------------------------------------------------------------------------

/// Called once when the audio pipeline is initialised.
- (void)audioProcessingInitializeWithSampleRate:(double)sampleRateHz
                                       channels:(int)channels {
    _sampleRateHz    = sampleRateHz;
    _numChannels     = channels;
    _samplesPerFrame = (int)(sampleRateHz / 100.0);  // 10 ms frame
    _holdFramesRemaining = 0;
    _smoothedGain        = 0.0f;
    [self recomputeCoefficients];
}

/// Called for every audio frame.
/// audioBuffer is an RTCAudioBuffer* — we access it via rawBufferForChannel:.
- (void)audioProcessingProcess:(id)audioBuffer {
    NoiseProcessingMode currentMode;
    float threshold, residualGain, attackMs, releaseMs;
    int   holdMs;

    @synchronized(self) {
        currentMode  = _mode;
        threshold    = _threshold;
        residualGain = _residualGain;
        attackMs     = _attackMs;
        releaseMs    = _releaseMs;
        holdMs       = _holdMs;
    }

    // disabled: passthrough.
    if (currentMode == NoiseProcessingModeDisabled) return;

    // denoise / hybrid not yet implemented — passthrough.
    if (currentMode == NoiseProcessingModeDenoise ||
        currentMode == NoiseProcessingModeHybrid) {
        // TODO(phase-2): run DeepFilterNet3 ONNX inference here.
        return;
    }

    // rmsGate (and vadGate falling back to rmsGate):

    // Recompute IIR coefficients from current ms values.
    // We do this inline (vs. caching) so that configure() changes take effect
    // on the very next frame without requiring a lock.
    float spf = (float)_samplesPerFrame;
    float sr  = (float)_sampleRateHz;

    float attackCoeff  = (attackMs  > 0.0f && sr > 0.0f)
        ? expf(-spf / (attackMs  * sr / 1000.0f)) : 0.0f;
    float releaseCoeff = (releaseMs > 0.0f && sr > 0.0f)
        ? expf(-spf / (releaseMs * sr / 1000.0f)) : 0.0f;

    // Get float PCM buffer for channel 0 via RTCAudioBuffer.
    // rawBufferForChannel: returns float* (non-owning).
    SEL rawBufSel = NSSelectorFromString(@"rawBufferForChannel:");
    if (![audioBuffer respondsToSelector:rawBufSel]) return;

    // RTCAudioBuffer.framesPerBand or num_frames — try both selector names.
    int n = 0;
    if ([audioBuffer respondsToSelector:NSSelectorFromString(@"framesPerBand")]) {
        n = (int)[(id)audioBuffer performSelector:
                  NSSelectorFromString(@"framesPerBand")];
    } else if ([audioBuffer respondsToSelector:NSSelectorFromString(@"numFrames")]) {
        n = (int)[(id)audioBuffer performSelector:
                  NSSelectorFromString(@"numFrames")];
    }
    if (n <= 0) return;

    // Retrieve the float buffer via IMP to avoid ARC bridging overhead and
    // to obtain the raw pointer safely.
    IMP imp = [audioBuffer methodForSelector:rawBufSel];
    typedef float* (*RawBufFunc)(id, SEL, int);
    float* ch0 = ((RawBufFunc)imp)(audioBuffer, rawBufSel, 0);
    if (!ch0) return;

    // ------------------------------------------------------------------
    // Compute RMS over channel 0.
    // ------------------------------------------------------------------
    float sumSq = 0.0f;
    for (int i = 0; i < n; ++i) {
        sumSq += ch0[i] * ch0[i];
    }
    float rms = sqrtf(sumSq / (float)n);
    @synchronized(self) { _rmsLevel = rms; }

    // ------------------------------------------------------------------
    // Gate logic with hold.
    // ------------------------------------------------------------------
    BOOL gateOpen;
    if (rms >= threshold) {
        int holdFrames = (sr > 0.0f)
            ? (int)((float)holdMs * sr / 1000.0f / spf)
            : 0;
        _holdFramesRemaining = holdFrames;
        gateOpen = YES;
    } else if (_holdFramesRemaining > 0) {
        _holdFramesRemaining--;
        gateOpen = YES;
    } else {
        gateOpen = NO;
    }

    // ------------------------------------------------------------------
    // IIR gain smoothing.
    // ------------------------------------------------------------------
    float targetGain = gateOpen ? 1.0f : residualGain;

    if (targetGain > _smoothedGain) {
        _smoothedGain =
            attackCoeff * _smoothedGain + (1.0f - attackCoeff) * targetGain;
    } else {
        _smoothedGain =
            releaseCoeff * _smoothedGain + (1.0f - releaseCoeff) * targetGain;
    }
    _smoothedGain = fmaxf(0.0f, fminf(1.0f, _smoothedGain));

    // ------------------------------------------------------------------
    // Apply gain to all channels in-place.
    // ------------------------------------------------------------------
    int numCh = 1;
    if ([audioBuffer respondsToSelector:NSSelectorFromString(@"numChannels")]) {
        numCh = (int)[(id)audioBuffer performSelector:
                      NSSelectorFromString(@"numChannels")];
    }

    for (int c = 0; c < numCh; ++c) {
        float* ch = ((RawBufFunc)imp)(audioBuffer, rawBufSel, c);
        if (!ch) continue;
        for (int i = 0; i < n; ++i) {
            ch[i] *= _smoothedGain;
        }
    }
}

/// Called when the audio pipeline is being torn down.
- (void)audioProcessingRelease {
    _sampleRateHz        = 0;
    _numChannels         = 0;
    _samplesPerFrame     = 0;
    _holdFramesRemaining = 0;
    _smoothedGain        = 0.0f;
    _attackCoeff         = 0.0f;
    _releaseCoeff        = 0.0f;
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

- (void)recomputeCoefficients {
    if (_sampleRateHz <= 0 || _samplesPerFrame <= 0) {
        _attackCoeff  = 0.0f;
        _releaseCoeff = 0.0f;
        return;
    }
    float spf = (float)_samplesPerFrame;
    float sr  = (float)_sampleRateHz;
    float atk, rel;
    @synchronized(self) {
        atk = _attackMs;
        rel = _releaseMs;
    }
    _attackCoeff  = (atk > 0.0f) ? expf(-spf / (atk * sr / 1000.0f)) : 0.0f;
    _releaseCoeff = (rel > 0.0f) ? expf(-spf / (rel * sr / 1000.0f)) : 0.0f;
}

@end
